File: src/image/Layer.java
package image;

import java.util.stream.Stream;

public class Layer {

	public static int MAX_OPACITY = 100;
	public static Coordinates DEFAULT_COORDINTES = new Coordinates();

	private String sourceFile;
	private long width;
	private long height;
	Pixel[][] pixelMatrix;
	boolean active;
	boolean visible;
	int opacity;
	int depth;
	Coordinates coordinates;
	Image image;

	// Constructor ------------------------------------------------------
	public Layer(Image image, final String filePath, long height, long width, boolean active, boolean visible, int opacity, final Coordinates coordinates, int depth) {
		this.sourceFile = filePath;
		this.height = height;
		this.width = width;
		this.pixelMatrix = new Pixel[(int) height][(int) width];
		this.active = active;
		this.visible = visible;
		this.opacity = opacity;
		this.coordinates = coordinates;
		this.depth = depth;
		this.image = image;
	}

	// Getters -----------------------------------------------------------
	public String getSourceFile() {
		return sourceFile;
	}

	public long getWidth() {
		return width;
	}

	public long getHeight() {
		return height;
	}

	public int getOpacity() {
		return opacity;
	}

	public boolean getVisibleStatus() {
		return visible;
	}

	public boolean getActiveStatus() {
		return active;
	}

	public long dataSize() {
		return width * height * 32;
	}

	public Pixel[][] getMatrix() {
		return this.pixelMatrix;
	}

	public Coordinates getCoordinates() {
		return coordinates;
	}

	// Setters -----------------------------------------------------------
	public void setOpacity(int opacity) {

		if (opacity < 0)
			this.opacity = 0;
		else
			this.opacity = opacity;

		Stream.of(this.pixelMatrix).forEach(row -> Stream.of(row).forEach(pixel -> {
			pixel.alpha = (byte) Math.min(255., ((double) opacity) / MAX_OPACITY * Byte.toUnsignedInt(pixel.alpha));

		}));
		this.image.setStatus(Image.Status.MODIFIED);
	}

	public int getDepth() {
		return this.depth;
	}
	public void setHeight(int height) {
		this.height = height;
	}

	public void setWidth(int width) {
		this.width = width;
	}


	public void setVisible(boolean newStatus) {
		this.visible = newStatus;
		this.image.setStatus(Image.Status.MODIFIED);

	}

	public void setActive(boolean newStatus) {
		this.active = newStatus;
	}

	public void setPosition(Coordinates coordinates) {
		this.coordinates = coordinates;
	}


	public void setDepth(int i) {
		this.depth = i;
	}

	public void setX(int x) {
		this.coordinates.x = x;
		this.image.setStatus(Image.Status.MODIFIED);

	}

	public void setY(int y) {
		this.coordinates.y = y;
		this.image.setStatus(Image.Status.MODIFIED);

	}
	// Operations -----------------------------------------------------------

	public void sharpen(int sharpIndex) {
		for (int i = 0; i < this.height; i += sharpIndex)
			for (int j = 0; j < this.width; j += sharpIndex) {
				final int m = Math.min((int) this.height, i + sharpIndex);
				final int n = Math.min((int) this.width, j + sharpIndex);
				for (int a = i; a < m; a++)
					for (int b = j; b < n; b++)
						this.pixelMatrix[a][b] = this.pixelMatrix[i][j];
			}
	}

	public Layer clone() {
		Layer newLayer = new Layer(this.image, this.sourceFile, this.height, this.width, this.active, this.visible, this.opacity, this.coordinates, this.depth);
		for (int i = 0; i < this.height; i++)
			for (int j = 0; j < this.width; j++)
				newLayer.pixelMatrix[i][j] = this.pixelMatrix[i][j].clone();
		return newLayer;
	}

}

File: C++ Project/src/layer.cpp
#include "layer.h"

#include <vector>


// Constructor ----------------------------------------------------------------
Layer::Layer(const std::string& _file_path, 
			unsigned long _height, 
			unsigned long _width, 
			bool _active,
			bool _visible,
			unsigned _opacity, 
			const Coordinates& _coordinates) :  source_file(_file_path),
												height(_height),
												width(_width),
												matrix(height, std::vector<Pixel>(width)),
												active(_active),
												visible(_visible),
												opacity(_opacity),
												coordinates(_coordinates) {}

// Cut ---------------------------------------------------------------
void Layer::cut(unsigned long x_axis_begin, unsigned long x_axis_end, unsigned long y_axis_begin, unsigned long y_axis_end) {
	matrix.erase(matrix.begin() + y_axis_begin, matrix.begin() + y_axis_end);		// cut rows

	if (x_axis_end - x_axis_begin)	// If there's anything to cut
		for (auto& vec : matrix)													// cut columns
			vec.erase(vec.begin() + x_axis_begin, vec.begin() + x_axis_end);
	
	this->width -= x_axis_end - x_axis_begin;
	this->height -= y_axis_end - y_axis_begin;
}

// Output ---------------------------------------------------------------

std::ostream& operator<<(std::ostream& os, const Layer& layer) {
	os << '\t' << "Source file: " << layer.source_file << std::endl;
	os << '\t' << "Dimensions: " << layer.width << " x " << layer.height << std::endl;
	os << '\t' << "Coordinates within image: (" << layer.coordinates.x << ", " << layer.coordinates.y << ")" << std::endl;
	os << '\t' << "Opacity: " << layer.opacity << std::endl;
	os << '\t' << "Visible: " << std::boolalpha << layer.visible << std::endl;
	os << '\t' << "Active: " << std::boolalpha << layer.active << std::endl;
	return os;
}

ptree Layer::to_ptree() const {
	ptree pt;
	pt.put("source_file", source_file);
	pt.put("width", width);
	pt.put("height", height);
	pt.put("coordinates", coordinates);
	pt.put("opacity", opacity);
	pt.put("visible", visible);
	pt.put("active", active);
	return pt;
}

// Opacity ---------------------------------------------------------------
void Layer::set_opacity(int _opacity) {
	if (_opacity < 0)
		opacity = 0;
	else
		opacity = _opacity;

	for (auto& vec : matrix)
		std::for_each(vec.begin(), vec.end(), [this](Pixel& pixel) { pixel.alpha = (Byte)std::min(255., (double)opacity / MAX_OPACITY * pixel.alpha); });
}
